File: client/typescript/src/sse.ts
export interface SSEFrame {
  /** Concatenated `data:` payload for this event. */
  data: string;
  /** Last `id:` field seen up to and including this frame, if any. */
  lastEventId?: string;
}

export async function* parseSSE(
  body: ReadableStream<Uint8Array>,
  signal?: AbortSignal,
): AsyncGenerator<SSEFrame, void, void> {
  const reader = body.getReader();
  const decoder = new TextDecoder("utf-8");
  let buffer = "";
  let lastEventId: string | undefined;

  const onAbort = () => {
    // cancel() rejects pending reads with the signal's reason, which
    // unwinds the generator via the for-await caller.
    reader.cancel(signal?.reason).catch(() => {});
  };
  signal?.addEventListener("abort", onAbort, { once: true });

  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      // Normalize line endings so a single split handles either CRLF or LF.
      buffer = buffer.replace(/\r\n/g, "\n");

      let sep: number;
      while ((sep = buffer.indexOf("\n\n")) !== -1) {
        const frame = buffer.slice(0, sep);
        buffer = buffer.slice(sep + 2);
        const data: string[] = [];
        for (const raw of frame.split("\n")) {
          if (raw === "" || raw.startsWith(":")) continue;
          const colon = raw.indexOf(":");
          const field = colon === -1 ? raw : raw.slice(0, colon);
          let value = colon === -1 ? "" : raw.slice(colon + 1);
          if (value.startsWith(" ")) value = value.slice(1);
          if (field === "data") data.push(value);
          else if (field === "id") lastEventId = value;
        }
        if (data.length === 0) continue;
        yield { data: data.join("\n"), lastEventId };
      }
    }
  } finally {
    signal?.removeEventListener("abort", onAbort);
    reader.releaseLock();
  }
}

File: client/python/src/hive/sse.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import AsyncGenerator, Optional

import httpx


@dataclass
class SSEFrame:
    data: str
    last_event_id: Optional[str] = field(default=None)


async def parse_sse(
    response: httpx.Response,
    abort: Optional[asyncio.Event] = None,
) -> AsyncGenerator[SSEFrame, None]:
    buffer = ""
    last_event_id: Optional[str] = None

    async for chunk in response.aiter_bytes():
        if abort is not None and abort.is_set():
            break

        buffer += chunk.decode("utf-8")
        buffer = buffer.replace("\r\n", "\n")

        while "\n\n" in buffer:
            if abort is not None and abort.is_set():
                return
            sep = buffer.index("\n\n")
            frame_text = buffer[:sep]
            buffer = buffer[sep + 2:]

            data_lines: list[str] = []
            for line in frame_text.split("\n"):
                if not line or line.startswith(":"):
                    continue
                colon = line.find(":")
                if colon == -1:
                    field_name, value = line, ""
                else:
                    field_name = line[:colon]
                    value = line[colon + 1:]
                    if value.startswith(" "):
                        value = value[1:]

                if field_name == "data":
                    data_lines.append(value)
                elif field_name == "id":
                    last_event_id = value

            if data_lines:
                yield SSEFrame(data="\n".join(data_lines), last_event_id=last_event_id)
